hollow/src/main/java/com/netflix/hollow/api/producer/HollowProducer.java

```diff
@@ -400,18 +452,92 @@ private void announce(HollowConsumer.ReadState readState) {
          *
          * @param blob the blob to publish
          * @param headerTags the header tags, in case these should be added as metadata on published artifacts.
+         * @return size of the blob published in bytes.
          */
-        public void publish(HollowProducer.Blob blob, Map<String, String> headerTags);
+        public abstract void publish(HollowProducer.Blob blob, Map<String, String> headerTags);
     }
 
-    public static interface Blob {
-        OutputStream newOutputStream();
-        InputStream newInputStream();
-        long getFromVersion();
-        long getToVersion();
-        Type getType();
-        void cleanup();
-        
+    public static class Blob {
+
+        protected final String namespace;
+        protected final long fromVersion;
+        protected final long toVersion;
+        protected final String dir;
+        protected final Blob.Type type;
+        protected final File file;
+
+        static Blob withNamespace(String namespace, long fromVersion, long toVersion, String dir, Blob.Type type) {
+            return new Blob(namespace, fromVersion, toVersion, dir, type);
+        }
+
+        private Blob(String namespace, long fromVersion, long toVersion, String dir, Blob.Type type) {
+            this.namespace = namespace;
+            this.fromVersion = fromVersion;
+            this.toVersion = toVersion;
+            this.type = type;
+            this.dir = dir;
+
+            switch (type) {
+                case SNAPSHOT:
```

Comment:
If you want, you can move methods onto the enum itself and replace switch with polymorphism.